src/os/bluestore/BlueStore.h

```diff
@@ -490,39 +511,190 @@ class BlueStore : public ObjectStore,
+    virtual int _rm_buffer(Buffer *b) = 0;
+    virtual void _adjust_buffer_size(Buffer *b, int64_t delta) = 0;
+    virtual void _touch_buffer(Buffer *b) = 0;
+
+    virtual void trim(uint64_t onode_max, uint64_t buffer_max) = 0;
+
+#ifdef DEBUG_CACHE
+    virtual void _audit(const char *s) = 0;
+#else
+    void _audit(const char *s) { /* no-op */ }
+#endif
+  };
+
+  /// simple LRU cache for onodes and buffers
+  struct LRUCache : public Cache {
+  private:
+    typedef boost::intrusive::list<
+      Onode,
+      boost::intrusive::member_hook<
+        Onode,
+	boost::intrusive::list_member_hook<>,
+	&Onode::lru_item> > onode_lru_list_t;
     typedef boost::intrusive::list<
       Buffer,
       boost::intrusive::member_hook<
 	Buffer,
 	boost::intrusive::list_member_hook<>,
 	&Buffer::lru_item> > buffer_lru_list_t;
+
+    onode_lru_list_t onode_lru;
+
+    buffer_lru_list_t buffer_lru;
+    uint64_t buffer_size = 0;
+
+  public:
+    void _add_onode(OnodeRef& o, int level) override {
+      if (level > 0)
+	onode_lru.push_front(*o);
+      else
+	onode_lru.push_back(*o);
+    }
+    void _rm_onode(OnodeRef& o) override {
+      auto q = onode_lru.iterator_to(*o);
+      onode_lru.erase(q);
+    }
+    void _touch_onode(OnodeRef& o) override;
+
+    void _add_buffer(Buffer *b, int level, Buffer *near) override {
+      if (near) {
+	auto q = buffer_lru.iterator_to(*near);
+	buffer_lru.insert(q, *b);
+      } else if (level > 0) {
+	buffer_lru.push_front(*b);
+      } else {
+	buffer_lru.push_back(*b);
+      }
+      buffer_size += b->length;
+    }
+    int _rm_buffer(Buffer *b) override {
+      buffer_size -= b->length;
+      auto q = buffer_lru.iterator_to(*b);
+      buffer_lru.erase(q);
+      return 1;
+    }
+    void _adjust_buffer_size(Buffer *b, int64_t delta) override {
+      buffer_size += delta;
+    }
+    void _touch_buffer(Buffer *b) override {
+      auto p = buffer_lru.iterator_to(*b);
+      buffer_lru.erase(p);
+      buffer_lru.push_front(*b);
+      _audit("_touch_buffer end");
+    }
+
+    void trim(uint64_t onode_max, uint64_t buffer_max) override;
+
+#ifdef DEBUG_CACHE
+    void _audit(const char *s) override;
+#endif
+  };
+
+  // 2Q cache for buffers, LRU for onodes
+  struct TwoQCache : public Cache {
+  private:
+    // stick with LRU for onodes for now (fixme?)
     typedef boost::intrusive::list<
       Onode,
       boost::intrusive::member_hook<
         Onode,
 	boost::intrusive::list_member_hook<>,
 	&Onode::lru_item> > onode_lru_list_t;
+    typedef boost::intrusive::list<
+      Buffer,
+      boost::intrusive::member_hook<
+	Buffer,
+	boost::intrusive::list_member_hook<>,
+	&Buffer::lru_item> > buffer_list_t;
 
-    std::mutex lock;                ///< protect lru and other structures
-    buffer_lru_list_t buffer_lru;
-    uint64_t buffer_size = 0;
     onode_lru_list_t onode_lru;
 
-    void _touch_onode(OnodeRef& o);
+    buffer_list_t buffer_hot;      //< "Am" hot buffers
```

Comment:
_add_buffer will insert if cache_private is already set to BUFFER_HOT.  This happens in BufferSpace::write() and did_read(), where cache_private gets the return value from _discard.  _discard returns the highest cache_private value of any buffers we are discarding (replacing).

The key to 2Q is that we get bumped to buffer_hot if we replace a buffer in the WARM_OUT (an empty buffer that we previously trimmed from warm_in) or a previously hot buffer.